src/main/java/io/fabric8/maven/docker/util/DockerFileUtil.java

```diff
@@ -78,40 +86,48 @@ public static String extractBaseImage(File dockerFile, Properties properties, St
      * Interpolate a docker file with the given properties and filter
      *
      * @param dockerFile docker file to interpolate
-     * @param properties properties to replace
-     * @param filter filter holding delimeters
+     * @param interpolator interpolator function for replacing properties
      * @return
      * @throws IOException
      */
-    public static String interpolate(File dockerFile, Properties properties, String filter) throws IOException {
+    public static String interpolate(File dockerFile, Function<String, String> interpolator) throws IOException {
         StringBuilder ret = new StringBuilder();
-        String[] delimiters = extractDelimiters(filter);
         try (BufferedReader reader = new BufferedReader(new FileReader(dockerFile))) {
             String line;
             while ((line = reader.readLine()) != null) {
-                ret.append(interpolateLine(line, properties, delimiters)).append(System.lineSeparator());
+                ret.append(interpolator.apply(line)).append(System.lineSeparator());
             }
         }
         return ret.toString();
     }
 
-    private static String interpolateLine(String line, Properties properties, String[] delimiters) {
-        if (delimiters == null || delimiters.length == 0) {
-            return line;
+    /**
+     * Create an interpolator function for properties with the given project and filter configuration.
+     *
+     * @param project The maven project.
+     * @param filter The filter configuration.
+     * @return An interpolator function for replacing properties.
+     */
+    public static Function<String, String> createInterpolator(MavenProject project, String filter) {
+        String[] delimiters = extractDelimiters(filter);
+        if (delimiters == null) {
+            return Functions.identity();
         }
-        Pattern propertyPattern =
-            Pattern.compile("(?<variable>" + Pattern.quote(delimiters[0]) + "(?<prop>.*?)" + Pattern.quote(delimiters[1]) + ")");
-        Matcher matcher = propertyPattern.matcher(line);
-        StringBuffer ret = new StringBuffer();
-        while (matcher.find()) {
-            String prop = matcher.group("prop");
-            String value = properties.containsKey(prop) ?
-                properties.getProperty(prop) :
-                matcher.group("variable");
-            matcher.appendReplacement(ret, value.replace("$","\\$"));
+        else {
```

Comment:
The `else` here seems useless, as you're returning in the other branch anyways. It does add a level of indentation though :)

Reply:
True.  Will nix it.